### PR TITLE
feat: DatadogAgent reconcileV2 skeleton

### PR DESCRIPTION
**Pull request by clamoriniere**
### What does this PR do?

* Define new reconcileV2() function that will be use with `v2alpha1.DatadogAgent`
* Add a new flag to the operator to switch to `reconcileV2`
* Create mostly the workflow for the Cluster-Agent component in `reconcileV2`.
* Create the skeleton of several function. they can be implemented in other PR.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

Write there any instructions and details you may have to test your PR.

### CODE REVIEW

File: controllers/datadogagent/controller_reconcile_v2.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+		logger.V(2).Info("Dependencies apply error", "errs", errs)
+		return result, errors.NewAggregate(errs)
+	}
+	// -----------------------------
+	// Start reconcile Components
+	// -----------------------------
+
+	newStatus := instance.Status.DeepCopy()
+
+	if requiredComponents.ClusterAgent.IsEnabled() {
```
Comment (vboulineau) [inline controllers/datadogagent/controller_reconcile_v2.go:112]: Is a component move from `Enabled` to `Disabled`, will it be clean-up itself by `store.Cleanup`?

File: controllers/datadogagent/component/clusteragent/default.go [hunk 2]
```diff
@@ -4,15 +4,28 @@ import (
...
+	"github.com/DataDog/datadog-operator/pkg/defaulting"
 )
 
+// NewDefaultClusterAgentDeployment return a new default cluster-agent deployment
+func NewDefaultClusterAgentDeployment(dda metav1.Object) *appsv1.Deployment {
+	deployment := component.NewDeployment(dda, apicommon.DefaultClusterAgentResourceSuffix, component.GetClusterAgentName(dda), component.GetClusterAgentVersion(dda), nil)
+
+	podTemplate := NewDefaultClusterAgentPodTemplateSpec(dda)
+	deployment.Spec.Template = *podTemplate
+	deployment.Spec.Replicas = apiutils.NewInt32Pointer(1)
```
Comment (celenechang) [inline controllers/datadogagent/component/clusteragent/default.go:24]: maybe should put in a constant

File: controllers/datadogagent/controller_reconcile_v2.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+func (r *Reconciler) updateStatusIfNeededV2(logger logr.Logger, agentdeployment *datadoghqv2alpha1.DatadogAgent, newStatus *datadoghqv2alpha1.DatadogAgentStatus, result reconcile.Result, currentError error) (reconcile.Result, error) {
+	// TODO(operator-ga): implement updateStatusIfNeededV2
+
+	// now := metav1.NewTime(time.Now())
+
+	return result, currentError
+}
+
+func (r *Reconciler) finalizeDadV2(reqLogger logr.Logger, obj client.Object) {
```
Comment (celenechang) [inline controllers/datadogagent/controller_reconcile_v2.go:156]: should it be `finalizeDDA`?

File: controllers/datadogagent/controller_reconcile_v2.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+			// Return and don't requeue
+			return result, nil
+		}
+		// Error reading the object - requeue the request.
+		return result, err
+	}
+
+	// check it the resource was properly decoded in v2
+	// if not it means it was a v1
+	if !apiequality.Semantic.DeepEqual(instance.Spec, datadoghqv2alpha1.DatadogAgentSpec{}) {
```
[the remark this answers is not in the document]
Comment (clamoriniere) [reply controllers/datadogagent/controller_reconcile_v2.go:43]: yes it was to support this case (that we discuss last time): not using the conversion webhook in this first version but call the convert function directly in the reconcile function

File: controllers/datadogagent/controller_reconcile_v2.go (hunk 1 above)
Comment (clamoriniere) [reply controllers/datadogagent/controller_reconcile_v2.go:112]: No, we need to build the Component cleanup. not done yet :)

File: controllers/datadogagent/component/clusteragent/default.go (hunk 2 above)
Comment (clamoriniere) [reply controllers/datadogagent/component/clusteragent/default.go:24]: agree

File: controllers/datadogagent/controller.go [hunk 5]
```diff
@@ -74,7 +75,15 @@ func NewReconciler(options ReconcilerOptions, client client.Client, versionInfo
 
 // Reconcile is similar to reconciler.Reconcile interface, but taking a context
 func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
-	resp, err := r.internalReconcile(ctx, request)
+	var resp reconcile.Result
+	var err error
+
+	if r.options.V2Enabled {
```
[the remark this answers is not in the document]
Comment (clamoriniere) [reply controllers/datadogagent/controller.go:81]: I added this hack, to allow me to link the new code pass, but keep the current `reconcile` function active to have all tests pass.

File: controllers/datadogagent/controller_reconcile_dca.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+	if alreadyExist {
+		// check if same hash
+		needUpdate := !comparison.IsSameSpecMD5Hash(hash, currentDeployment.GetAnnotations())
+		if !needUpdate {
+			// no need to update to stop here the process
+			return reconcile.Result{}, nil
+		}
+
+		logger.Info("Updating an existing Cluster Agent Deployment", "deployment.Namespace", deployment.Namespace, "deployment.Name", deployment.Name, "currentHash", hash)
```
[the remark this answers is not in the document]
Comment (clamoriniere) [reply controllers/datadogagent/controller_reconcile_dca.go:89]: I used it a lot with wpa logs with custom column

File: controllers/datadogagent/controller_reconcile_v2.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (vboulineau) [reply controllers/datadogagent/controller_reconcile_v2.go:43]: Alright but in this case it requires another "top-level" reconciler there: https://github.com/DataDog/datadog-operator/blob/main/controllers/datadogagent_controller.go#L222 with `v2alpha1`

[the remark this answers is not in the document]
Comment (clamoriniere) [reply controllers/datadogagent/controller_reconcile_v2.go:43]: indeed. As I said this PR doesn't bring the full implementation working. It is here to help to // other function development.
I will work on supporting both version by looking at the CRD version

File: controllers/datadogagent/component/clusteragent/default.go (hunk 2 above)
Comment (clamoriniere) [reply controllers/datadogagent/component/clusteragent/default.go]: done in https://github.com/DataDog/datadog-operator/pull/491/commits/c61514c7b095598bd22bc5959a00946c3df99f82

File: controllers/datadogagent/controller_reconcile_v2.go (hunk 3 above)
Comment (celenechang) [reply controllers/datadogagent/controller_reconcile_v2.go:156]: oh, ok

File: controllers/datadogagent/component/new.go [hunk 7]
```diff
@@ -3,18 +3,84 @@ package component
...
+
+	edsv1alpha1 "github.com/DataDog/extendeddaemonset/api/v1alpha1"
 )
 
 // NewDeployment use to generate the skeleton of a new deployment based on few information
 func NewDeployment(owner metav1.Object, componentKind, componentName, version string, selector *metav1.LabelSelector) *appsv1.Deployment {
-	labels := object.GetDefaultLabels(owner, componentName, version)
-	labels[apicommon.AgentDeploymentNameLabelKey] = owner.GetName()
-	labels[apicommon.AgentDeploymentComponentLabelKey] = componentKind
+	labels, annotations, selector := getdefaultInfo(owner, componentKind, componentName, version, selector)
```
Comment (celenechang) [inline controllers/datadogagent/component/new.go:16]: suggestion, `getDefaultMetadata` ?

File: controllers/datadogagent/feature/enabledefault/feature.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+func (f *defaultFeature) Configure(dda *v2alpha1.DatadogAgent) feature.RequiredComponents {
+	trueValue := true
+	return feature.RequiredComponents{
+		ClusterAgent: feature.RequiredComponent{
+			IsRequired: &trueValue,
+		},
+		Agent: feature.RequiredComponent{
+			IsRequired: &trueValue,
+		},
+	}
```
Comment (celenechang) [inline controllers/datadogagent/feature/enabledefault/feature.go:36]: so this "feature" ensures that the agent and cluster agent are enabled if nothing else is set?

File: controllers/datadogagent/feature/kubernetesstatecore/feature.go [hunk 9]
```diff
@@ -79,10 +79,15 @@ func (f *ksmFeature) Configure(dda *v2alpha1.DatadogAgent) feature.RequiredCompo
 		}
 	}
 
-	return feature.RequiredComponents{
-		ClusterAgent:       feature.RequiredComponent{IsRequired: &f.enable},
-		ClusterCheckRunner: feature.RequiredComponent{IsRequired: &f.clusterChecksEnabled},
+	output := feature.RequiredComponents{}
+	if f.enable {
```
Comment (celenechang) [inline controllers/datadogagent/feature/kubernetesstatecore/feature.go:83]: could we nest this (as well as the `if` block starting on line 72) within the first `if` block?

File: controllers/datadogagent/feature/enabledefault/feature.go (hunk 8 above)
Comment (clamoriniere) [reply controllers/datadogagent/feature/enabledefault/feature.go:36]: yes,
I will add in another PR the creation of the default associate resources: serviceAccount, condigmap, RBAC...

File: controllers/datadogagent/feature/kubernetesstatecore/feature.go [hunk 10]
```diff
@@ -59,47 +59,49 @@ type ksmFeature struct {
...
-	if dda.Spec.Features.ClusterChecks != nil && apiutils.BoolValue(dda.Spec.Features.ClusterChecks.Enabled) {
-		f.clusterChecksEnabled = true
-		if apiutils.BoolValue(dda.Spec.Features.ClusterChecks.UseClusterChecksRunners) {
-			f.rbacSuffix = common.CheckRunnersSuffix
-			f.serviceAccountName = v2alpha1.GetClusterChecksRunnerServiceAccount(dda)
-		} else {
-			f.serviceAccountName = v2alpha1.GetClusterAgentServiceAccount(dda)
+		if dda.Spec.Features.ClusterChecks != nil && apiutils.BoolValue(dda.Spec.Features.ClusterChecks.Enabled) {
+			f.clusterChecksEnabled = true
+			output.ClusterCheckRunner.IsRequired = &f.clusterChecksEnabled
```
Comment (celenechang) [inline controllers/datadogagent/feature/kubernetesstatecore/feature.go:76]: should this be set to `true` under `if apiutils.BoolValue(dda.Spec.Features.ClusterChecks.UseClusterChecksRunners)` ?

**Review by celenechang — APPROVED** [review]
added a question, but it can be addressed in a subsequent PR

File: controllers/datadogagent/feature/kubernetesstatecore/feature.go (hunk 10 above)
Comment (clamoriniere) [reply controllers/datadogagent/feature/kubernetesstatecore/feature.go:76]: right fixing now

Comment (clamoriniere) [reply controllers/datadogagent/feature/kubernetesstatecore/feature.go:76]: @celenechang 
I made the fix in https://github.com/DataDog/datadog-operator/pull/496

**Pull request merged**